File: redundant-connection/redundant-connection.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        UnionFind uf = new UnionFind(n + 1);
        for (int[] e : edges) {
            if (!uf.union(e[0], e[1])) return e;
        }
        return new int[]{};
    }
}

class UnionFind {
    int[] parent, size;
    public UnionFind(int n) {
        parent = new int[n]; size = new int[n];
        for (int i = 1; i < n; i++) {
            parent[i] = i; size[i] = 1;
        }
    }

    public int find(int x) {
        if (x == parent[x]) return x;
        return find(parent[x]);
    }

    public boolean union(int u, int v) {
        int parentU = find(u), parentV = find(v);
        if (parentU == parentV) return false;
        if (size[parentU] > size[parentV]) {
            size[parentU] += size[parentV];
            parent[parentV] = parentU;
        } else {
            size[parentV] += size[parentU];
            parent[parentU] = parentV;
        }
        return true;
    }
}

File: 684-redundant-connection/redundant-connection.py
class UnionFind:
    def __init__(self, n):
        self.rank = [0] * n
        self.parent = list(range(n))
    def find(self, a):
        if self.parent[a] != a:
            self.parent[a] = self.find(self.parent[a])
        return self.parent[a]
    def union(self, a, b):
        parent_of_a = self.find(a)
        parent_of_b = self.find(b)
        if parent_of_a == parent_of_b:
            return False
        if self.rank[parent_of_a] < self.rank[parent_of_b]:
            self.parent[parent_of_a] = parent_of_b
        elif self.rank[parent_of_a] > self.rank[parent_of_b]:
            self.parent[parent_of_b] = parent_of_a
        else:
            self.parent[parent_of_b] = parent_of_a
            self.rank[parent_of_a] += 1
        return True
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        uf = UnionFind(n)
        for edge in edges:
            if not uf.union(edge[0] - 1, edge[1] - 1):
                return edge
        return []